photos/db.go

```diff
@@ -0,0 +1,478 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Spencer Kimball (spencer.kimball@gmail.com)
+
+package main
+
+import (
+	"database/sql"
+	"errors"
+	"math/rand"
+	"net/url"
+	"time"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+	_ "github.com/lib/pq"
+)
+
+const (
+	lgSchema = `
+CREATE DATABASE IF NOT EXISTS photos;
+
+CREATE TABLE IF NOT EXISTS photos.users (
+  id           INT,
+  photoCount   INT,
+  commentCount INT,
+  name         STRING,
+  address      STRING,
+
+  PRIMARY KEY (id)
+);
+
+CREATE TABLE IF NOT EXISTS photos.photos (
+  id           BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
+  userID       INT,
+  commentCount INT,
+  caption      STRING,
+  latitude     FLOAT,
+  longitude    FLOAT,
+  timestamp    TIMESTAMP,
+
+  PRIMARY KEY (id)
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS photosByUserID ON photos.photos (userID, id);
+
+CREATE TABLE IF NOT EXISTS photos.comments (
+  photoID   BYTES,
+  commentID BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
+  userID    INT,
+  message   STRING,
+  timestamp TIMESTAMP,
+
+  PRIMARY KEY (photoID, commentID)
+);
+`
+)
+
+// pgURL returns a Postgresql URL string for the specified context.
+func pgURL(ctx Context) string {
+	certUser := security.RootUser
+	options := url.Values{}
+	options.Add("sslmode", "verify-full")
+	options.Add("sslcert", security.ClientCertPath(ctx.CertsDir, certUser))
+	options.Add("sslkey", security.ClientKeyPath(ctx.CertsDir, certUser))
+	options.Add("sslrootcert", security.CACertPath(ctx.CertsDir))
+	url := url.URL{
+		Scheme:   "postgres",
+		User:     url.User(certUser),
+		Host:     ctx.DBAddr,
+		RawQuery: options.Encode(),
+	}
+	return url.String()
+}
+
+// openDB opens the database connection according to the context.
+func openDB(ctx Context) (*sql.DB, error) {
+	return sql.Open("postgres", pgURL(ctx))
+}
+
+// initSchema creates the database schema if it doesn't exist.
+func initSchema(db *sql.DB) error {
+	_, err := db.Exec(lgSchema)
+	return err
+}
+
+// dropDatabase drops the database.
+func dropDatabase(db *sql.DB) error {
+	_, err := db.Exec("DROP DATABASE photos;")
```

Comment:
Do you want an error here if the `photos` database doesn't exist? Probably better to `DROP DATABASE IF EXISTS photos`.